Fix for config_section bug with subcommands

tbartlett0 opened the pull request:
* Since pushing that bunch of changes didn't look to work so well, I'll just push my own change on its own for now... *

Reorder code in cmdline:create_parser to correct a bug where the config_section setting was being ignored for the main function when subcommands are used. (The subcommand setup is overwriting the config section used by populate_parser, so do it after setting up the main function.)

ropez committed:
Use file log format when stdout is not a tty

tbartlett0 committed:
Fix for config_section bug with subcommands

Reorder code in cmdline:create_parser to correct a bug where the config_section setting was being ignored for the main function when subcommands are used. (The subcommand setup is overwriting the config section used by populate_parser, so do it after setting up the main function.)

Lapin0t committed:
feature: simplier flag creation

Lapin0t committed:
Merge pull request #1 from sh0oki/master

When submodules are defined, require a submodule.

Lapin0t committed:
exit code is value returned by last callback

Lapin0t committed:
change 'no_no' to 'double_flags'

esc committed:
fix urls

sh0oki commented:
@tbartlett0 Thanks! 
Can you please check why the Python 2.6 test fails?

sh0oki commented on begin/cmdline.py:
```diff
@@ -171,6 +171,15 @@ def create_parser(func, env_prefix=None, config_file=None, config_section=None,
+    while hasattr(func, '__wrapped__') and not hasattr(func, '__signature__'):
+        if isinstance(func, extensions.Extension):
+            func.add_arguments(parser, defaults)
+            have_extensions = True
```
Is this variable still needed?

tbartlett0 replied:
It doesn't look like it, this and the other just below are the only uses I see in the entire code. Both assignments, and never used elsewhere.

tbartlett0 committed:
Attempting to merge changes
Merge https://github.com/sh0oki/begins into test

tbartlett0 committed:
Merge https://github.com/Lapin0t/begins into test

Conflicts:
	begin/cmdline.py

tbartlett0 committed:
Merge branch 'fix_pypi_badge' of https://github.com/esc/begins into test

tbartlett0 committed:
Merge https://github.com/snaptv/begins into test
Default to file log format when stdout is not a tty

tbartlett0 committed:
Fixed merge conflict code that I'd accidentally missed

tbartlett0 committed:
Revert "exit code is value returned by last callback"
This change removed the ability to use 'if begin.start():' so
isn't backward compatible: discussion needed before merging.

This reverts commit 318db8a44cfba6c9fc1601b09b0d06f128358ef5.

tbartlett0 commented:
The Python 2.6 failure appears to be something to do with _coverage_:
```
Traceback (most recent call last):
  File "/home/travis/virtualenv/python2.6.9/bin/coverage", line 9, in <module>
    load_entry_point('coverage==4.3.2', 'console_scripts', 'coverage')()
< ... >
  File "/home/travis/virtualenv/python2.6.9/lib/python2.6/site-packages/coverage/misc.py", line 287, in <module>
    class StopEverything(unittest.SkipTest):

AttributeError: 'module' object has no attribute 'SkipTest'
```
I don't _think_ anything I changed should have caused this? I will see if I can replicate it on my own box with python 2.6.

tbartlett0 commented:
I had a painful time trying to get python 2.6 to even install and work, but once I did I was able to run the tests without error. Is it possible that something in the test environment has changed since the last successful 2.6 build?

Update: from https://docs.python.org/2/library/unittest.html#skipping-tests-and-expected-failures
"25.3.6. Skipping tests and expected failures¶
_New in version 2.7._"
So unittest.SkipTest doesn't exist in python 2.6, but why is the test trying to use it and why did it not have this issue previously? (https://travis-ci.org/sh0oki/begins/jobs/189842536)

My test VM installed coverage 4.3.4, which contains in `misc.py`:
```
    287 # unittest.SkipTest doesn't exist in Python 2.6. We backport it with
    288 # unittest2 in the coverage.py test suite.  But for production, we don't need
    289 # to derive from SkipTest, so if it doesn't exist, just use Exception.
    290 
    291 class StopEverything(getattr(unittest, 'SkipTest', Exception)):
```

tbartlett0 commented:
Oops, committing the other forks' changes back to Master has apparently updated the pull request. (On the bright side, it seems my fix for the test failure has worked, and even py2.6 is passing now.)

If you're happy to merge all these patches then please do, otherwise let me know and I can roll back my master to just my change and submit the others separately.

tbartlett0 commented:
I've deleted 'have_extensions' per your comment as it's never used.

tbartlett0 committed:
Correcting test failure for flag parameters.
Adjusted test to expect the non-default flag option to be first.
Adjusted help for flag options to only display for non-default option,
as per sh0oki's existing code's function.

tbartlett0 committed:
Remove unused have_extensions variable.

sh0oki commented:
@tbartlett0 Thanks!
Can you please commit in this pull request your changes related to the specific issue they are fixing?
I agree there are other good changes to merge, but let's merge them in another pull request.
WDYT?

tbartlett0 commented:
Ok. This has got too messy, and as far as I can tell reverting the patches I pulled would be even more messy, so I'm going to make a new pull request on a separate branch with only my changes.
